### DISCUSSION TITLE
Selection 'best' face to use from set of same face extractions

### DISCUSSION THREAD

**Discussion by rleaver152** [policratus/pupyl#136] [Ideas]
When generating a set of faces from photos,  you can end up with many faces of the same person.   It would be useful to be able to rank these in terms of 'best fit' for future face recognition within a corpus.   So you dont have to search by n faces of the same reference person,  just reduce to one template face for each person,  which is somehow 'the best' to use for that person.  Perhaps there is a way to configure a pupyl search to deliver this?

**Comment by policratus** [comment]
As always, another great idea @rleaver152! `pupyl` is primarily an image search platform and it does that in the more general way possible (no "knowledge" of classes, but similarities). Implementing face recognition (a subject that I also love and have some projects regarding this matter) is a feature that should increase drastically the cool functionalities of `pupyl`.

At the end of the day, `pupyl` will be expanding its search features (to be honest, I got myself several times wanting that `pupyl` also search by face, not only general similarities). Hence, to start working with this, I suggest a flow (because face recog. is different from image search):

1. The search process happens as always, with preprocessing steps;
2. Adding a really fast step to detect faces on the images being indexed (should not impact the search process) will help change the flow: if a face is detected, `pupyl` can choose if it will extract face features or stay on the image features extraction;
3. From now on, the set of algorithms for search must change (if a face was detected).

There's more on this discussion, however. For instance:

1. How we deal with mixed modes (faces and similarity), on the same database?
2. Detecting faces during the index process will reduce the extraction speed?

Thanks again for this great idea!

**Comment by rleaver152** [comment]
Hi - sorry - I didn't mean actually perform face recognition with Pupyl,   but somehow be able to configure it to deliver a face which is an optimal similarity with a group of faces of the same person,  yet optimally dissimilar to the rest of groups of faces of different people.  At the moment, you can generate n ranked similar images beautifully.   I am just trying to get my head around if you could somehow configure the same algorithms to provide the 'best' single match amongst known 'same' people whilst maximising dissimilarity between the rest of 'not same' people.          This is still dealing with image similarities of course, but just that every image is simply a face,  already framed and registered from other programs to present Pupyl with a face database ready to process.

**Reply by rleaver152** [reply]
as a ps,  I found this wonderful repo which I've adjusted to output the "best" face from a sorted ranked list of its peers.     https://github.com/deepcam-cn/FaceQuality

**Reply by policratus** [reply]
Hey @rleaver152, sorry for the late response. I became a father and time have been really scarce right now.

Hmmmm, got it, I guess. So instead of doing face recognition, we use the same algorithms already implemented, but if all indexed images are faces, a face query image will obviously return faces, but may return a _rank_ or _factor_ (based on [FaceQuality](https:// github.com/deepcam-cn/FaceQuality), the probability distribution between 0 and 1).

That's it?

**Comment by rleaver152** [comment]
Hi,  congratulations!!

Yes,   basically to try to only provide the ‘best’ faces for subsequent database use in face recognition rather than a massive set only some of which might actually be contributory (or so near duplicates eg rotations, that they are superfluous and wasteful of resource in face rec later on)
